File: src/assembler/rules-assembler.ts
/**
 * RulesAssembler — assembles .claude/rules/ and skills/ from source knowledge packs.
 *
 * @remarks
 * Migrated from Python `assembler/rules_assembler.py` (540 lines).
 * Split into three modules for 250-line compliance:
 * - `rules-assembler.ts` — main class (this file)
 * - `rules-identity.ts` — identity content builders
 * - `rules-conditionals.ts` — conditional assembly functions
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import { getActiveRoutes } from "../domain/core-kp-routing.js";
import { getStackPackName } from "../domain/stack-pack-mapping.js";
import { findVersionDir } from "../domain/version-resolver.js";
import { auditRulesContext } from "./auditor.js";
import { buildIdentityContent, fallbackDomainContent } from "./rules-identity.js";
import {
  copyDatabaseRefs,
  copyCacheRefs,
  assembleSecurityRules,
  assembleCloudKnowledge,
  assembleInfraKnowledge,
} from "./rules-conditionals.js";

/** Result returned by {@link RulesAssembler.assemble}. */
export interface AssembleResult {
  readonly files: string[];
  readonly warnings: string[];
}

/** Assembles .claude/rules/ and skills/ from source knowledge packs. */
export class RulesAssembler {
  /** Orchestrate all assembly layers. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    engine: TemplateEngine,
  ): AssembleResult {
    const rulesDir = path.join(outputDir, "rules");
    const skillsDir = path.join(outputDir, "skills");
    fs.mkdirSync(rulesDir, { recursive: true });
    fs.mkdirSync(skillsDir, { recursive: true });
    const generated: string[] = [];
    generated.push(...this.copyCoreRules(resourcesDir, rulesDir, engine));
    generated.push(...this.routeCoreToKps(config, resourcesDir, skillsDir));
    generated.push(...this.copyLanguageKps(config, resourcesDir, skillsDir));
    generated.push(...this.copyFrameworkKps(config, resourcesDir, skillsDir));
    generated.push(this.generateProjectIdentity(config, rulesDir));
    generated.push(this.copyDomainTemplate(config, resourcesDir, rulesDir, engine));
    generated.push(...copyDatabaseRefs(config, resourcesDir, skillsDir, engine));
    generated.push(...copyCacheRefs(config, resourcesDir, skillsDir));
    generated.push(...assembleSecurityRules(config, resourcesDir, skillsDir));
    generated.push(...assembleCloudKnowledge(config, resourcesDir, skillsDir));
    generated.push(...assembleInfraKnowledge(config, resourcesDir, skillsDir));
    const audit = auditRulesContext(rulesDir);
    return { files: generated, warnings: [...audit.warnings] };
  }

  /** Layer 1: Copy core-rules/*.md with placeholder replacement. */
  private copyCoreRules(
    resourcesDir: string,
    rulesDir: string,
    engine: TemplateEngine,
  ): string[] {
    const coreRules = path.join(resourcesDir, "core-rules");
    if (!fs.existsSync(coreRules) || !fs.statSync(coreRules).isDirectory()) {
      return [];
    }
    const generated: string[] = [];
    const files = fs.readdirSync(coreRules).filter((f) => f.endsWith(".md")).sort();
    for (const file of files) {
      const content = fs.readFileSync(path.join(coreRules, file), "utf-8");
      const replaced = engine.replacePlaceholders(content);
      const dest = path.join(rulesDir, file);
      fs.writeFileSync(dest, replaced, "utf-8");
      generated.push(dest);
    }
    return generated;
  }

  /** Layer 1b: Route core detailed rules to knowledge packs. */
  private routeCoreToKps(
    config: ProjectConfig,
    resourcesDir: string,
    skillsDir: string,
  ): string[] {
    const coreDir = path.join(resourcesDir, "core");
    if (!fs.existsSync(coreDir) || !fs.statSync(coreDir).isDirectory()) {
      return [];
    }
    const routes = getActiveRoutes(config);
    const generated: string[] = [];
    for (const route of routes) {
      const src = path.join(coreDir, route.sourceFile);
      if (!fs.existsSync(src) || !fs.statSync(src).isFile()) continue;
      const destDir = path.join(skillsDir, route.kpName, "references");
      fs.mkdirSync(destDir, { recursive: true });
      const dest = path.join(destDir, route.destFile);
      fs.copyFileSync(src, dest);
      generated.push(dest);
    }
    return generated;
  }

  /** Layer 2: Route language files to coding-standards and testing KPs. */
  private copyLanguageKps(
    config: ProjectConfig,
    resourcesDir: string,
    skillsDir: string,
  ): string[] {
    const lang = config.language.name;
    const langDir = path.join(resourcesDir, "languages", lang);
    if (!fs.existsSync(langDir) || !fs.statSync(langDir).isDirectory()) {
      return [];
    }
    const codingRefs = path.join(skillsDir, "coding-standards", "references");
    const testingRefs = path.join(skillsDir, "testing", "references");
    fs.mkdirSync(codingRefs, { recursive: true });
    fs.mkdirSync(testingRefs, { recursive: true });
    const generated: string[] = [];
    generated.push(...this.copyLangCommon(langDir, codingRefs, testingRefs));
    generated.push(...this.copyLangVersion(config, langDir, codingRefs));
    return generated;
  }

  private copyLangCommon(
    langDir: string,
    codingRefs: string,
    testingRefs: string,
  ): string[] {
    const common = path.join(langDir, "common");
    if (!fs.existsSync(common) || !fs.statSync(common).isDirectory()) {
      return [];
    }
    const generated: string[] = [];
    const files = fs.readdirSync(common).filter((f) => f.endsWith(".md")).sort();
    for (const file of files) {
      const dest = file.includes("testing") ? testingRefs : codingRefs;
      const target = path.join(dest, file);
      fs.copyFileSync(path.join(common, file), target);
      generated.push(target);
    }
    return generated;
  }

  private copyLangVersion(
    config: ProjectConfig,
    langDir: string,
    codingRefs: string,
  ): string[] {
    const versionDir = findVersionDir(
      langDir, config.language.name, config.language.version,
    );
    if (!versionDir) return [];
    const generated: string[] = [];
    const files = fs.readdirSync(versionDir).filter((f) => f.endsWith(".md")).sort();
    for (const file of files) {
      const target = path.join(codingRefs, file);
      fs.copyFileSync(path.join(versionDir, file), target);
      generated.push(target);
    }
    return generated;
  }

  /** Layer 3: Route framework files to stack-patterns KP. */
  private copyFrameworkKps(
    config: ProjectConfig,
    resourcesDir: string,
    skillsDir: string,
  ): string[] {
    const fw = config.framework.name;
    const packName = getStackPackName(fw);
    if (!packName) return [];
    const fwDir = path.join(resourcesDir, "frameworks", fw);
    if (!fs.existsSync(fwDir) || !fs.statSync(fwDir).isDirectory()) {
      return [];
    }
    const refsDir = path.join(skillsDir, packName, "references");
    fs.mkdirSync(refsDir, { recursive: true });
    const generated: string[] = [];
    generated.push(...this.copyFwCommon(fwDir, refsDir));
    generated.push(...this.copyFwVersion(config, fwDir, refsDir));
    return generated;
  }

  private copyFwCommon(fwDir: string, refsDir: string): string[] {
    const common = path.join(fwDir, "common");
    if (!fs.existsSync(common) || !fs.statSync(common).isDirectory()) {
      return [];
    }
    const generated: string[] = [];
    const files = fs.readdirSync(common).filter((f) => f.endsWith(".md")).sort();
    for (const file of files) {
      const target = path.join(refsDir, file);
      fs.copyFileSync(path.join(common, file), target);
      generated.push(target);
    }
    return generated;
  }

  private copyFwVersion(
    config: ProjectConfig,
    fwDir: string,
    refsDir: string,
  ): string[] {
    const versionDir = findVersionDir(
      fwDir, config.framework.name, config.framework.version,
    );
    if (!versionDir) return [];
    const generated: string[] = [];
    const files = fs.readdirSync(versionDir).filter((f) => f.endsWith(".md")).sort();
    for (const file of files) {
      const target = path.join(refsDir, file);
      fs.copyFileSync(path.join(versionDir, file), target);
      generated.push(target);
    }
    return generated;
  }

  /** Layer 4: Generate 01-project-identity.md. */
  private generateProjectIdentity(
    config: ProjectConfig,
    rulesDir: string,
  ): string {
    const dest = path.join(rulesDir, "01-project-identity.md");
    const content = buildIdentityContent(config);
    fs.writeFileSync(dest, content, "utf-8");
    return dest;
  }

  /** Layer 4: Copy/generate 02-domain.md. */
  private copyDomainTemplate(
    config: ProjectConfig,
    resourcesDir: string,
    rulesDir: string,
    engine: TemplateEngine,
  ): string {
    const dest = path.join(rulesDir, "02-domain.md");
    const template = path.join(resourcesDir, "templates", "domain-template.md");
    if (fs.existsSync(template) && fs.statSync(template).isFile()) {
      const content = fs.readFileSync(template, "utf-8");
      const replaced = engine.replacePlaceholders(content);
      fs.writeFileSync(dest, replaced, "utf-8");
    } else {
      fs.writeFileSync(dest, fallbackDomainContent(config), "utf-8");
    }
    return dest;
  }
}

File: java/src/main/java/dev/iadev/assembler/RulesAssembler.java
package dev.iadev.assembler;

import dev.iadev.domain.stack.CoreKpRouting;
import dev.iadev.domain.stack.StackPackMapping;
import dev.iadev.domain.stack.VersionResolver;
import dev.iadev.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * Assembles {@code .claude/rules/} and {@code .claude/skills/}
 * from source knowledge packs and templates.
 *
 * <p>This is the first assembler in the pipeline (position 1 of
 * 23 per RULE-005). It generates 5 core rules unconditionally
 * plus conditional rules based on project features.</p>
 *
 * <p>Assembly layers:
 * <ol>
 *   <li>Core rules — copy core-rules/*.md with placeholder
 *       replacement</li>
 *   <li>Core KP routing — route core docs to knowledge
 *       packs</li>
 *   <li>Language KPs — language-specific coding standards
 *       and testing conventions</li>
 *   <li>Framework KPs — framework-specific patterns</li>
 *   <li>Project identity — generate 01-project-identity.md
 *       (overwrites template copy)</li>
 *   <li>Domain template — generate/copy 02-domain.md</li>
 *   <li>Conditionals — database, cache, security, cloud,
 *       infrastructure</li>
 * </ol>
 *
 * <p>Example usage:
 * <pre>{@code
 * Assembler rules = new RulesAssembler();
 * List<String> files = rules.assemble(
 *     config, engine, outputDir);
 * }</pre>
 * </p>
 *
 * @see Assembler
 * @see RulesIdentity
 * @see RulesConditionals
 */
public final class RulesAssembler implements Assembler {

    private final Path resourcesDir;

    /**
     * Creates a RulesAssembler using classpath resources.
     */
    public RulesAssembler() {
        this(resolveClasspathResources());
    }

    /**
     * Creates a RulesAssembler with an explicit resources
     * directory.
     *
     * @param resourcesDir the base resources directory
     */
    public RulesAssembler(Path resourcesDir) {
        this.resourcesDir = resourcesDir;
    }

    /**
     * {@inheritDoc}
     *
     * <p>Orchestrates all assembly layers, returning the
     * list of generated file paths.</p>
     */
    @Override
    public List<String> assemble(
            ProjectConfig config,
            TemplateEngine engine,
            Path outputDir) {
        Path rulesDir = outputDir.resolve("rules");
        Path skillsDir = outputDir.resolve("skills");
        CopyHelpers.ensureDirectory(rulesDir);
        CopyHelpers.ensureDirectory(skillsDir);

        Map<String, Object> context =
                buildContext(config);

        List<String> generated = new ArrayList<>();
        generated.addAll(
                copyCoreRules(rulesDir, engine, context));
        generated.addAll(
                routeCoreToKps(config, skillsDir));
        generated.addAll(
                copyLanguageKps(config, skillsDir));
        generated.addAll(
                copyFrameworkKps(config, skillsDir));
        generated.add(
                generateProjectIdentity(config, rulesDir));
        generated.add(
                copyDomainTemplate(
                        config, rulesDir, engine, context));
        generated.addAll(
                copyConditionals(
                        config, skillsDir, engine, context));

        return generated;
    }

    /**
     * Builds the placeholder context map from config.
     *
     * @param config the project configuration
     * @return context map for placeholder replacement
     */
    static Map<String, Object> buildContext(
            ProjectConfig config) {
        return Map.ofEntries(
                Map.entry("project_name",
                        config.project().name()),
                Map.entry("project_purpose",
                        config.project().purpose()),
                Map.entry("language_name",
                        config.language().name()),
                Map.entry("language_version",
                        config.language().version()),
                Map.entry("framework_name",
                        config.framework().name()),
                Map.entry("framework_version",
                        config.framework().version()),
                Map.entry("build_tool",
                        config.framework().buildTool()),
                Map.entry("architecture_style",
                        config.architecture().style()),
                Map.entry("domain_driven",
                        String.valueOf(
                                config.architecture()
                                        .domainDriven())),
                Map.entry("event_driven",
                        String.valueOf(
                                config.architecture()
                                        .eventDriven())),
                Map.entry("container",
                        config.infrastructure().container()),
                Map.entry("orchestrator",
                        config.infrastructure()
                                .orchestrator()),
                Map.entry("database_name",
                        config.data().database().name()),
                Map.entry("cache_name",
                        config.data().cache().name()));
    }

    private List<String> copyCoreRules(
            Path rulesDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        Path coreRules =
                resourcesDir.resolve("core-rules");
        if (!Files.exists(coreRules)
                || !Files.isDirectory(coreRules)) {
            return List.of();
        }
        List<String> generated = new ArrayList<>();
        List<Path> files = listMdFilesSorted(coreRules);
        for (Path file : files) {
            String dest = CopyHelpers.copyTemplateFile(
                    file,
                    rulesDir.resolve(
                            file.getFileName().toString()),
                    engine, context);
            generated.add(dest);
        }
        return generated;
    }

    private List<String> routeCoreToKps(
            ProjectConfig config, Path skillsDir) {
        Path coreDir = resourcesDir.resolve("core");
        if (!Files.exists(coreDir)
                || !Files.isDirectory(coreDir)) {
            return List.of();
        }
        var routes =
                CoreKpRouting.getActiveRoutes(config);
        List<String> generated = new ArrayList<>();
        for (var route : routes) {
            Path src = coreDir.resolve(route.sourceFile());
            if (!Files.exists(src)
                    || !Files.isRegularFile(src)) {
                continue;
            }
            Path destDir = skillsDir.resolve(
                    route.kpName() + "/references");
            CopyHelpers.ensureDirectory(destDir);
            Path dest = destDir.resolve(route.destFile());
            generated.add(
                    CopyHelpers.copyStaticFile(src, dest));
        }
        return generated;
    }

    private List<String> copyLanguageKps(
            ProjectConfig config, Path skillsDir) {
        String lang = config.language().name();
        Path langDir =
                resourcesDir.resolve("languages/" + lang);
        if (!Files.exists(langDir)
                || !Files.isDirectory(langDir)) {
            return List.of();
        }
        Path codingRefs = skillsDir.resolve(
                "coding-standards/references");
        Path testingRefs = skillsDir.resolve(
                "testing/references");
        CopyHelpers.ensureDirectory(codingRefs);
        CopyHelpers.ensureDirectory(testingRefs);
        List<String> generated = new ArrayList<>();
        generated.addAll(
                copyLangCommon(langDir,
                        codingRefs, testingRefs));
        generated.addAll(
                copyLangVersion(config,
                        langDir, codingRefs));
        return generated;
    }

    private List<String> copyLangCommon(
            Path langDir,
            Path codingRefs, Path testingRefs) {
        Path common = langDir.resolve("common");
        if (!Files.exists(common)
                || !Files.isDirectory(common)) {
            return List.of();
        }
        List<String> generated = new ArrayList<>();
        List<Path> files = listMdFilesSorted(common);
        for (Path file : files) {
            String name =
                    file.getFileName().toString();
            Path dest = name.contains("testing")
                    ? testingRefs : codingRefs;
            generated.add(
                    CopyHelpers.copyStaticFile(
                            file,
                            dest.resolve(name)));
        }
        return generated;
    }

    private List<String> copyLangVersion(
            ProjectConfig config,
            Path langDir, Path codingRefs) {
        Optional<Path> versionDir =
                VersionResolver.findVersionDir(
                        langDir,
                        config.language().name(),
                        config.language().version());
        if (versionDir.isEmpty()) {
            return List.of();
        }
        List<String> generated = new ArrayList<>();
        List<Path> files =
                listMdFilesSorted(versionDir.get());
        for (Path file : files) {
            generated.add(
                    CopyHelpers.copyStaticFile(
                            file,
                            codingRefs.resolve(
                                    file.getFileName()
                                            .toString())));
        }
        return generated;
    }

    private List<String> copyFrameworkKps(
            ProjectConfig config, Path skillsDir) {
        String fw = config.framework().name();
        String packName =
                StackPackMapping.getStackPackName(fw);
        if (packName.isEmpty()) {
            return List.of();
        }
        Path fwDir =
                resourcesDir.resolve("frameworks/" + fw);
        if (!Files.exists(fwDir)
                || !Files.isDirectory(fwDir)) {
            return List.of();
        }
        Path refsDir = skillsDir.resolve(
                packName + "/references");
        CopyHelpers.ensureDirectory(refsDir);
        List<String> generated = new ArrayList<>();
        generated.addAll(copyFwCommon(fwDir, refsDir));
        generated.addAll(
                copyFwVersion(config, fwDir, refsDir));
        return generated;
    }

    private List<String> copyFwCommon(
            Path fwDir, Path refsDir) {
        Path common = fwDir.resolve("common");
        if (!Files.exists(common)
                || !Files.isDirectory(common)) {
            return List.of();
        }
        List<String> generated = new ArrayList<>();
        List<Path> files = listMdFilesSorted(common);
        for (Path file : files) {
            generated.add(
                    CopyHelpers.copyStaticFile(
                            file,
                            refsDir.resolve(
                                    file.getFileName()
                                            .toString())));
        }
        return generated;
    }

    private List<String> copyFwVersion(
            ProjectConfig config,
            Path fwDir, Path refsDir) {
        Optional<Path> versionDir =
                VersionResolver.findVersionDir(
                        fwDir,
                        config.framework().name(),
                        config.framework().version());
        if (versionDir.isEmpty()) {
            return List.of();
        }
        List<String> generated = new ArrayList<>();
        List<Path> files =
                listMdFilesSorted(versionDir.get());
        for (Path file : files) {
            generated.add(
                    CopyHelpers.copyStaticFile(
                            file,
                            refsDir.resolve(
                                    file.getFileName()
                                            .toString())));
        }
        return generated;
    }

    private String generateProjectIdentity(
            ProjectConfig config, Path rulesDir) {
        Path dest =
                rulesDir.resolve("01-project-identity.md");
        String content =
                RulesIdentity.buildContent(config);
        try {
            Files.writeString(
                    dest, content, StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new UncheckedIOException(
                    "Failed to write identity rule", e);
        }
        return dest.toString();
    }

    private String copyDomainTemplate(
            ProjectConfig config,
            Path rulesDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        Path dest = rulesDir.resolve("02-domain.md");
        Path template = resourcesDir.resolve(
                "templates/domain-template.md");
        if (Files.exists(template)
                && Files.isRegularFile(template)) {
            return CopyHelpers.copyTemplateFile(
                    template, dest, engine, context);
        }
        try {
            Files.writeString(dest,
                    RulesIdentity.fallbackDomainContent(
                            config),
                    StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new UncheckedIOException(
                    "Failed to write domain rule", e);
        }
        return dest.toString();
    }

    private List<String> copyConditionals(
            ProjectConfig config,
            Path skillsDir,
            TemplateEngine engine,
            Map<String, Object> context) {
        List<String> generated = new ArrayList<>();
        generated.addAll(
                RulesConditionals.copyDatabaseRefs(
                        config, resourcesDir,
                        skillsDir, engine, context));
        generated.addAll(
                RulesConditionals.copyCacheRefs(
                        config, resourcesDir, skillsDir));
        generated.addAll(
                RulesConditionals.assembleSecurityRules(
                        config, resourcesDir, skillsDir));
        generated.addAll(
                RulesConditionals.assembleCloudKnowledge(
                        config, resourcesDir, skillsDir));
        generated.addAll(
                RulesConditionals.assembleInfraKnowledge(
                        config, resourcesDir, skillsDir));
        return generated;
    }

    private static List<Path> listMdFilesSorted(Path dir) {
        try (Stream<Path> stream = Files.list(dir)) {
            return stream
                    .filter(f -> f.toString()
                            .endsWith(".md"))
                    .filter(Files::isRegularFile)
                    .sorted()
                    .toList();
        } catch (IOException e) {
            throw new UncheckedIOException(
                    "Failed to list directory: " + dir, e);
        }
    }

    private static Path resolveClasspathResources() {
        var url = RulesAssembler.class.getClassLoader()
                .getResource("core-rules");
        if (url == null) {
            return Path.of("src/main/resources");
        }
        return Path.of(url.getPath()).getParent();
    }
}
